src/txmempool.cpp

```diff
@@ -89,34 +334,106 @@ void CTxMemPool::AddTransactionsUpdated(unsigned int n)
+
+    const CTransaction& tx = newit->GetTx();
+    std::set<uint256> setParentTransactions;
+    for (unsigned int i = 0; i < tx.vin.size(); i++) {
         mapNextTx[tx.vin[i].prevout] = CInPoint(&tx, i);
+        setParentTransactions.insert(tx.vin[i].prevout.hash);
+    }
+    // Don't bother worrying about child transactions of this one.
+    // Normal case of a new transaction arriving is that there can't be any
+    // children, because such children would be orphans.
+    // An exception to that is if a transaction enters that used to be in a block.
+    // In that case, our disconnect block logic will call UpdateTransactionsFromBlock
+    // to clean up the mess we're leaving here.
+
+    // Update ancestors with information about this tx
+    BOOST_FOREACH (const uint256 &phash, setParentTransactions) {
+        txiter pit = mapTx.find(phash);
+        if (pit != mapTx.end()) {
+            UpdateParent(newit, pit, true);
+        }
+    }
+    UpdateAncestorsOf(true, hash, setAncestors);
+
     nTransactionsUpdated++;
     totalTxSize += entry.GetTxSize();
-    cachedInnerUsage += entry.DynamicMemoryUsage();
     minerPolicyEstimator->processTransaction(entry, fCurrentEstimate);
 
     return true;
 }
 
+// TODO: replace this hash with an iterator?
+void CTxMemPool::removeUnchecked(const uint256& hash)
+{
+    indexed_transaction_set::iterator it = mapTx.find(hash);
+
+    BOOST_FOREACH(const CTxIn& txin, it->GetTx().vin)
+        mapNextTx.erase(txin.prevout);
+
+    totalTxSize -= it->GetTxSize();
+    cachedInnerUsage -= it->DynamicMemoryUsage();
+    cachedInnerUsage -= memusage::DynamicUsage(mapLinks[it].parents) + memusage::DynamicUsage(mapLinks[it].children);
+    mapLinks.erase(it);
+    mapTx.erase(it);
+    nTransactionsUpdated++;
+    minerPolicyEstimator->removeTx(hash);
+}
+
+// Calculates descendants of hash that are not already in setDescendants, and adds to 
+// setDescendants. Assumes hash is already a tx in the mempool and setMemPoolChildren
+// is correct for tx and all descendants.
+// Also assumes that if an entry is in setDescendants already, then all
+// in-mempool descendants of it are already in setDescendants as well, so that we
+// can save time by not iterating over those entries.
+void CTxMemPool::CalculateDescendants(const uint256 &hash, std::set<uint256> &setDescendants)
+{
+    std::set<uint256> stage;
+    if (setDescendants.count(hash) == 0) {
+        stage.insert(hash);
+    }
+    // Traverse down the children of each hash, only adding children that are not
+    // accounted for in setDescendants already (because those children have either
+    // already been walked, or will be walked in this iteration).
+    while (!stage.empty()) {
+        setDescendants.insert(stage.begin(), stage.end());
+        std::set<uint256> setNext;
+        BOOST_FOREACH(const uint256 &stagehash, stage) {
+            indexed_transaction_set::iterator it = mapTx.find(stagehash);
```

Comment:
Check/assert `it == mapTx.end()`?
Would only be possible on the first `stagehash` AFAIK